core/course_controller/CourseController.py

```diff
@@ -1,48 +1,226 @@
+            raise ValueError("Course code is required for saving lab sections.")
+
+        if not course_data.lab_sections_codes:
+            # If no lab sections are provided, return early
+            return "No lab sections to add."
+
+        try:
+            course = Course.objects.get(course_code=course_data.course_code)
+        except Course.DoesNotExist:
+            raise ValueError("Course with the given code does not exist.")
+
+        existing_sections = LabSection.objects.filter(course=course).values_list('lab_section_number', flat=True)
+
+        for section_number in course_data.lab_sections_codes:
+            if section_number in existing_sections:
+                raise ValueError(f"Duplicate lab section number: {section_number}")
+            LabSection.objects.create(
+                course=course,
+                lab_section_number=section_number,
+                start_time="00:00",
+                end_time="00:00",
+            )
+
+        return f"Lab sections added to course {course_data.course_code}"
+
+    @staticmethod
+    def save_course_sections(course_data: CourseFormData) -> str:
+        """
+        Helper method for save course to handle adding new course sections.
+
+        Pre-conditions: course_data is valid in the CourseFormData provided (course_code is provided).
+        Post-conditions: Adds record(s) to the CourseSection table with the relation to the course_code provided in CourseFormData.
+        Side-effects: new record(s) added to the CourseSection table.
+        Returns: list of id(s) of the new course section objects added to the table or an IllegalValueException if CourseFormData
+            doesn't have a course_code.
         """
-        pass
\ No newline at end of file
+        if not course_data.course_code:
+            raise ValueError("Course code is required for saving course sections.")
+
+        if not course_data.instructor:
+            raise ValueError("Instructor is required for saving course sections.")
+
+        if not course_data.course_sections_codes:
+            # If no course sections are provided, return early
+            return "No course sections to add."
+
+        try:
+            course = Course.objects.get(course_code=course_data.course_code)
+        except Course.DoesNotExist:
+            raise ValueError("Course with the given code does not exist.")
+
+        try:
+            instructor = User.objects.get(username=course_data.instructor.username, role="Instructor")
+        except User.DoesNotExist:
+            raise ValueError("The provided instructor does not exist or is not an Instructor.")
+
+        existing_sections = CourseSection.objects.filter(course=course).values_list('course_section_number', flat=True)
+
+        for section_number in course_data.course_sections_codes:
+            if section_number in existing_sections:
+                raise ValueError(f"Duplicate course section number: {section_number}")
```

Comment:
If you create a course, then go to edit it again and click save it will try to save the course sections again. This method should delete old ones, create new ones, but not touch existing ones.

Reply:
Might have overlooked this in the tests